File: programs/private-perps/src/instructions/initialize_market.rs
use anchor_lang::prelude::*;
use crate::{state::*, InitializeMarket};

pub fn handler(ctx: Context<InitializeMarket>, params: InitializeMarketParams) -> Result<()> {
    let market = &mut ctx.accounts.market;
    market.admin = ctx.accounts.admin.key();
    market.symbol = params.symbol;
    market.collateral_mint = params.collateral_mint;
    market.fee_receiver = params.fee_receiver;
    market.oracle = params.oracle;
    market.mxe_authority = params.mxe_authority;
    market.maker_fee_bps = params.maker_fee_bps;
    market.taker_fee_bps = params.taker_fee_bps;
    market.liquidation_fee_bps = params.liquidation_fee_bps;
    market.initial_margin_bps = params.initial_margin_bps;
    market.maintenance_margin_bps = params.maintenance_margin_bps;
    market.max_leverage = params.max_leverage;
    market.funding_epoch_secs = params.funding_epoch_secs;
    market.last_funding_ts = Clock::get()?.unix_timestamp;
    market.paused = false;
    market.bump = ctx.bumps.market;
    let registry = &mut ctx.accounts.registry;
    if registry.authority == Pubkey::default() {
        registry.authority = ctx.accounts.admin.key();
        registry.bump = ctx.bumps.registry;
    }
    registry.market_count += 1;
    let pv = &mut ctx.accounts.protocol_vault;
    pv.market = market.key();
    pv.collateral_mint = params.collateral_mint;
    pv.bump = ctx.bumps.protocol_vault;
    emit!(MarketInitialized { market: market.key(), symbol: params.symbol, admin: ctx.accounts.admin.key() });
    Ok(())
}

#[event]
pub struct MarketInitialized {
    pub market: Pubkey,
    pub symbol: [u8; 8],
    pub admin: Pubkey,
}

File: scripts/initialize_market.ts
import * as anchor from "@coral-xyz/anchor";
import { PublicKey, SystemProgram, SYSVAR_RENT_PUBKEY } from "@solana/web3.js";
import { TOKEN_PROGRAM_ID } from "@solana/spl-token";
import * as fs from "fs";
import * as path from "path";

const PROGRAM_ID = new PublicKey("Bn8G8L4egZaL1LeWx2QZRFSRVWJWL8dEkj35i392tUmJ");
const USDC_MINT = new PublicKey("4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU");
const PYTH_BTC_ORACLE = new PublicKey("HovQMDrbAgApt5TNnqKe9HNFVnk22GPCuXRYbKRMC9i");
const SYMBOL = [66, 84, 67, 45, 80, 69, 82, 80];

async function main() {
  const idl = JSON.parse(fs.readFileSync(path.join(__dirname, "../target/idl/private_perps.json"), "utf8"));
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);
  const admin = provider.wallet.publicKey;
  const program = new anchor.Program(idl as any, provider);

  const [registry] = PublicKey.findProgramAddressSync([Buffer.from("registry")], PROGRAM_ID);
  const symBuf = Buffer.from(SYMBOL);
  const [market] = PublicKey.findProgramAddressSync([Buffer.from("market"), symBuf], PROGRAM_ID);
  const [protocolVaultAta] = PublicKey.findProgramAddressSync([Buffer.from("vault"), market.toBuffer()], PROGRAM_ID);
  const [protocolVaultAuthority] = PublicKey.findProgramAddressSync([Buffer.from("vault_auth"), market.toBuffer()], PROGRAM_ID);
  const [protocolVault] = PublicKey.findProgramAddressSync([Buffer.from("protocol_vault"), market.toBuffer()], PROGRAM_ID);

  console.log("Admin:        ", admin.toBase58());
  console.log("Market PDA:   ", market.toBase58());
  console.log("Registry PDA: ", registry.toBase58());

  const params = {
    symbol: SYMBOL,
    collateralMint: USDC_MINT,
    feeReceiver: admin,
    oracle: PYTH_BTC_ORACLE,
    mxeAuthority: admin,
    makerFeeBps: -5,
    takerFeeBps: 10,
    liquidationFeeBps: 50,
    initialMarginBps: 1000,
    maintenanceMarginBps: 500,
    maxLeverage: 10,
    fundingEpochSecs: new anchor.BN(3600),
  };

  const tx = await program.methods
    .initializeMarket(params)
    .accounts({
      admin,
      registry,
      market,
      collateralMint: USDC_MINT,
      protocolVaultAta,
      protocolVaultAuthority,
      protocolVault,
      tokenProgram: TOKEN_PROGRAM_ID,
      systemProgram: SystemProgram.programId,
      rent: SYSVAR_RENT_PUBKEY,
    })
    .rpc();

  console.log("✅ Market initialized! TX:", tx);
  console.log("Explorer: https://explorer.solana.com/tx/" + tx + "?cluster=devnet");
  console.log("Market PDA:", market.toBase58());
}

main().catch(err => { console.error(err); process.exit(1); });
